### PR TITLE
Fix for issue #46 - lang_tester runs into a loop on MacOS.

### PR DESCRIPTION
**Pull request by ummarikar**
This is an in-progress PR to show the steps taken to solve issue #46.

### CODE REVIEW

**Comment by ummarikar** [comment]
Screenshot 2019-12-08 at 23 53 23
This is the current error I'm getting my fix was to switch (back) to using the following lines:
https://github.com/softdevteam/lang_tester/blob/a1f51ec8b747cb252a0b8df4846ff43c8c0c6c8b/src/tester.rs#L661-L662
Pushing the changes now.

**Comment by ummarikar** [comment]
This solves the error I was getting however now the tests take an incredibly long time to run and when they finally show up they fail. I had to cancel the run however as it was taking too long.

**Comment by ummarikar** [comment]
In my local version of `filedescriptor` I switched from macOS's implementation of poll(2) which is a wrapper around select(2) to using the actual poll(2) on macOS (obviously this is not an option and was just done for experimentation purposes). The result was that the yksom tests actually run and don't take an incredibly long time but they all fail. And, it looks like the expected output is being added to stderr:
Screenshot 2019-12-09 at 00 22 17
Looking into this now.

**Comment by ummarikar** [comment]
Tests now all pass when using the macOS `poll(2)` in `filedescriptor` but using the `select(2)` implementation produces the same problem I was getting before where the tests wouldn't run.

**Comment by ltratt** [comment]
OK, let's break this PR down into some stages and start again (it will probably be easier to close this and open a new PR, but that's up to you):

  1) Use the suggestion from https://github.com/wez/wzsh/pull/1 for converting pipes into `FileDescriptor` objects. That's a commit on its own.

  2) The child stdin/stderr file descriptors need to be marked as non-blocking. I don't know if the `filedescriptor` library allows that. If not, you'll need to use `fcntl`. This should be a commit on its own. There is a tiny chance that this might worm around the `poll` bug on OS X, at which point job done! If not, well, it needed doing anyway.

  3) If `poll` still doesn't work, then use old fashioned `select` instead. This will require a bit more work.

**Comment by ummarikar** [comment]
I'll keep using this PR then we can squash everything where appropriate when we're done :)

**Comment by ummarikar** [comment]
Non-blocking implemented. I'm wondering if I have done this correctly. I want to be sure I have before I move on to trying the `select(2)` route.

File: src/tester.rs [hunk 1]
```diff
@@ -651,8 +652,14 @@ fn run_cmd(
         .spawn()
         .unwrap_or_else(|_| fatal(&format!("Couldn't run command {:?}.", cmd)));
 
-    let mut stderr = FileDescriptor::dup(child.stderr.as_ref().unwrap()).unwrap();
     let mut stdout = FileDescriptor::dup(child.stdout.as_ref().unwrap()).unwrap();
+    let mut stderr = FileDescriptor::dup(child.stderr.as_ref().unwrap()).unwrap();
```
Comment (ltratt) [inline src/tester.rs]: I'm not sure why this line has moved.

File: src/tester.rs [hunk 2]
```diff
@@ -651,8 +652,14 @@ fn run_cmd(
...
+
+    fcntl(stdout.as_socket_descriptor(), F_SETFL(OFlag::O_NONBLOCK))
```
Comment (ltratt) [inline src/tester.rs]: I'm pretty sure that you need to do `flags = fcntl(..., F_GETFL, 0)` to read the flags, then `fcntl(..., F_SETFL, flags | O_NONBLOCK)`.

**Pull request closed without merging**